Cannot compile with thread affinity enabled, undefined reference to `sched_getcpu'

I have tried to compile both the latest release and the latest git code, but both fail to compile if I enable thread affinity.
I am pretty sure I really need to have thread affinity working, because on a 4 socket Barcelona cluster (4*12=48 cores per node) I am getting absolutely abysmal scaling on DGELSY, sometimes even heavy negative scaling. I am fairly sure this is due to having 4 NUMA regions, and thread affinity could help resolve it. But I cannot get openblas to compile with thread affinity on this cluster.
The cluster is using GCC 4.7.2.

Here is the part where failures happen:

> gfortran -O2 -Wall -m64   -o dblat1 dblat1.o ../libopenblas_barcelonap-r0.2.20.dev.a -lm -lpthread -lgfortran -lm -lpthread -lgfortran
> sblat2.f: In function ‘schk1’:
> sblat2.f:646:0: warning: ‘nargs’ may be used uninitialized in this function [-Wmaybe-uninitialized]
> dblat2.f: In function ‘dchk1’:
> dblat2.f:646:0: warning: ‘nargs’ may be used uninitialized in this function [-Wmaybe-uninitialized]
> ../libopenblas_barcelonap-r0.2.20.dev.a(init.o): In function 'gotoblas_affinity_init':
> init.c:(.text+0xd07): undefined reference to 'sched_getcpu'
> collect2: error: ld returned 1 exit status
> make[1]: *** [sblat1] Error 1
> make[1]: *** Waiting for unfinished jobs....

--- a/driver/others/init.c
+++ b/driver/others/init.c
@@ -354,6 +354,24 @@ static int numa_check(void) {
   return common -> num_nodes;
 }
 
+#if defined(__GLIBC_PREREQ)
+#if !__GLIBC_PREREQ(2, 6)
+int sched_getcpu(void)
+{
+int cpu;
+FILE *fp = NULL;
+if ( (fp = fopen("/proc/self/stat", "r")) == NULL)
+   return -1;
+if ( fscanf( fp, "%*s%*s%*s%*s%*s%*s%*s%*s%*s%*s%*s%*s%*s%*s%*s%*s%*s%*s%*s%*s%*s%*s%*s%*s%*s%*s%*s%*s%*s%*s%*s%*s%*s%*s%*s%*s%*s%*s%d", &cpu) != 1) {
+  fclose (fp);
+  return -1;
+  }
+  fclose (fp);
+  return(cpu);
+}
+#endif
+#endif
+
 static void numa_mapping(void) {
 
   int node, cpu, core;
@@ -808,7 +826,6 @@ void gotoblas_affinity_init(void) {
   common -> shmid = pshmid;
 
   if (common -> magic != SH_MAGIC) {
-
 #ifdef DEBUG
     fprintf(stderr, "Shared Memory Initialization.\n");
 #endif
@@ -830,7 +847,7 @@ void gotoblas_affinity_init(void) {
     if (common -> num_nodes > 1) numa_mapping();
 
     common -> final_num_procs = 0;
-    for(i = 0; i < common -> avail_count; i++) common -> final_num_procs += rcount(common -> avail[i]) + 1;   //Make the max cpu number.
+    for(i = 0; i < common -> avail_count; i++) common -> final_num_procs += rcount(common -> avail[i]) + 1;   //Make the max cpu number. 
 
     for (cpu = 0; cpu < common -> final_num_procs; cpu ++) common -> cpu_use[cpu] =  0;
 

